sql/src/main/java/io/crate/execution/engine/collect/RemoteCollectorFactory.java

```diff
@@ -98,73 +99,64 @@ public RemoteCollectorFactory(ClusterService clusterService,
                                               RoutedCollectPhase collectPhase,
                                               CollectTask collectTask,
                                               ShardCollectorProviderFactory shardCollectorProviderFactory) {
-        final UUID childJobId = UUID.randomUUID(); // new job because subContexts can't be merged into an existing job
-
-        Collector<Row, ?, List<Object[]>> listCollector = Collectors.mapping(Row::materialize, Collectors.toList());
-        CollectingRowConsumer<?, List<Object[]>> consumer = new CollectingRowConsumer<>(listCollector);
-
-        ShardStateAwareRemoteCollector shardStateAwareRemoteCollector = new ShardStateAwareRemoteCollector(
-            shardId,
-            consumer,
-            clusterService,
-            indicesService,
-            getLocalCollectorProvider(shardCollectorProviderFactory, collectPhase, collectTask, consumer),
-            getRemoteCollectorProvider(childJobId, shardId, collectPhase, collectTask, consumer),
-            searchTp,
-            threadPool.getThreadContext());
-
+        ShardStateObserver shardStateObserver = new ShardStateObserver(clusterService, threadPool.getThreadContext());
+        CompletableFuture<ShardRouting> shardBecameActive = shardStateObserver.waitForActiveShard(shardId);
+        Runnable onClose = () -> {};
+        Consumer<Throwable> kill = killReason -> {
+            shardBecameActive.cancel(true);
+            shardBecameActive.completeExceptionally(killReason);
+        };
         return CollectingBatchIterator.newInstance(
-            () -> {},
-            shardStateAwareRemoteCollector::kill,
-            () -> {
-                shardStateAwareRemoteCollector.doCollect();
-                return consumer.completionFuture().thenApply(results -> results.stream().map(Buckets.arrayToSharedRow())::iterator);
-            },
+            onClose,
+            kill,
+            () -> shardBecameActive.thenCompose(
+                activePrimaryRouting -> retrieveRows(activePrimaryRouting, collectPhase, collectTask, shardCollectorProviderFactory)),
             true
         );
     }
 
-    private Function<IndexShard, BatchIterator<Row>> getLocalCollectorProvider(ShardCollectorProviderFactory shardCollectorProviderFactory,
-                                                                               RoutedCollectPhase collectPhase,
-                                                                               CollectTask collectTask,
-                                                                               RowConsumer consumer) {
-        return indexShard -> {
+    private CompletableFuture<List<Row>> retrieveRows(ShardRouting activePrimaryRouting,
```

Comment:
imho, it was a bit more readable and easier to grasp the logic immediately when we had `getLocalCollectorProvider` and `getRemoteCollectorProvider` methods

Reply:
Any suggestions how to restore that? The if/else blocks within `retrieveRows` are rather short, so I'm not sure if it makes sense to extract some methods to clarify that one case is remote and another local

Reply:
thought of keeping those methods, but it doesn't really matter 👍